tests/conftest.py

```diff
@@ -34,11 +52,71 @@ def mock_config_entry() -> MockConfigEntry:
     )
 
 
+@pytest.fixture
+def hass_ws_client(
+    aiohttp_client: ClientSessionGenerator,
+    hass_access_token: str,
+    hass: HomeAssistant,
+    socket_enabled: None,
+) -> WebSocketGenerator:
+    """Websocket client fixture connected to websocket server."""
+
+    async def create_client(
+        hass: HomeAssistant = hass, access_token: str | None = hass_access_token
+    ) -> MockHAClientWebSocket:
+        """Create a websocket client."""
+        assert await async_setup_component(hass, "websocket_api", {})
+        client = await aiohttp_client(hass.http.app)
+        websocket = await client.ws_connect(URL)
+        auth_resp = await websocket.receive_json()
+        assert auth_resp["type"] == TYPE_AUTH_REQUIRED
+
+        if access_token is None:
+            await websocket.send_json({"type": TYPE_AUTH, "access_token": "incorrect"})
+        else:
+            await websocket.send_json({"type": TYPE_AUTH, "access_token": access_token})
+
+        auth_ok = await websocket.receive_json()
+        assert auth_ok["type"] == TYPE_AUTH_OK
+
+        def _get_next_id() -> Generator[int]:
+            i = 0
+            while True:
+                yield (i := i + 1)
+
+        id_generator = _get_next_id()
+
+        def _send_json_auto_id(data: dict[str, Any]) -> Coroutine[Any, Any, None]:
+            data["id"] = next(id_generator)
+            return websocket.send_json(data)
+
+        async def _remove_device(device_id: str, config_entry_id: str) -> Any:
+            await _send_json_auto_id(
+                {
+                    "type": "config/device_registry/remove_config_entry",
+                    "config_entry_id": config_entry_id,
+                    "device_id": device_id,
+                }
+            )
+            return await websocket.receive_json()
+
+        # wrap in client
+        wrapped_websocket = cast(MockHAClientWebSocket, websocket)
+        wrapped_websocket.client = client
+        wrapped_websocket.send_json_auto_id = _send_json_auto_id
+        wrapped_websocket.remove_device = _remove_device
+        return wrapped_websocket
+
+    return create_client
+
+
 # Copied from homeassistant:
 # https://github.com/home-assistant/core/blob/2f121874987b5f19aed6b5769b9880c5322d95d0/tests/components/command_line/__init__.py#L9
 @contextmanager
 def mock_asyncio_subprocess_run(
-    response: bytes = b"", returncode: int = 0, exception: Exception | None = None
+    responses: bytes = iter([b""]),
```

Comment:
I'd like to find a better implementation for this, but since I was calling `create_subprocess_exec` twice in the `async_list_backups` function I needed to mock two separate return values.